java/org/contikios/cooja/Cooja.java

```diff
@@ -2390,34 +2381,75 @@ public void actionPerformed(ActionEvent e) {
       progressDialog = null;
     }
 
-    // Load simulation in this thread, while showing progress monitor
-    boolean shouldRetry;
-    Simulation newSim = null;
-    do {
-      try {
-        shouldRetry = false;
-        cooja.doRemoveSimulation(false);
-        PROGRESS_WARNINGS.clear();
-        newSim = loadSimulationConfig(configFile, quick, rewriteCsc, manualRandomSeed);
+    final var cfgFile = configFile;
+    // SwingWorker can pass information from worker to process() through publish().
+    // Communicate information the other way through this shared queue.
+    final var channel = new SynchronousQueue<Integer>(true);
+    var worker = new SwingWorker<Simulation, SimulationCreationException>() {
+      @Override
+      public Simulation doInBackground() {
+        boolean shouldRetry;
+        Simulation newSim = null;
+        do {
+          try {
+            shouldRetry = false;
+            cooja.doRemoveSimulation(false);
+            PROGRESS_WARNINGS.clear();
+            newSim = loadSimulationConfig(cfgFile, quick, rewriteCsc, manualRandomSeed);
+          } catch (SimulationCreationException e) {
+            publish(e);
+            try {
+              shouldRetry = channel.take() == 1;
+            } catch (InterruptedException ex) {
+              cooja.doRemoveSimulation(false);
+              return null;
+            }
+          }
+        } while (shouldRetry);
+        return newSim;
+      }
 
-        /* Optionally show compilation warnings */
-        boolean hideWarn = Boolean.parseBoolean(
-            Cooja.getExternalToolsSetting("HIDE_WARNINGS", "false")
-        );
+      @Override
+      protected void process(List<SimulationCreationException> exs) {
+        for (var e : exs) {
+          var retry = showErrorDialog(Cooja.getTopParentContainer(), "Simulation load error", e, true);
+          try {
+            channel.put(retry ? 1 : 0);
+          } catch (InterruptedException ex) {
+            cooja.doRemoveSimulation(false);
```

Comment:
What do you think about cancelling the worker instead? It should cause the worker thread to be interrupted if it is still waiting for the channel?

Reply:
This is the code defining the worker, so the compiler complains that the variable might not have been initialized.

Reply:
Strange, I did not seem to get any warnings when I tested.
```
          try {
              channel.put(retry ? 1 : 0);
          } catch (InterruptedException ex) {
              cancel(true);
          }
          break;
```

Reply:
I thought the cancel worked on the current thread, but if that cancels the worker thread it works. Updated the patch.